File: src/main/java/gargoyle/ct/pref/impl/prop/CTPrefEnumProperty.java
package gargoyle.ct.pref.impl.prop;

import gargoyle.ct.convert.impl.EnumConverter;
import gargoyle.ct.pref.CTPreferencesProvider;
import org.jetbrains.annotations.NotNull;

public class CTPrefEnumProperty<E extends Enum<E>> extends CTPrefProperty<E> {
    public CTPrefEnumProperty(@NotNull CTPreferencesProvider provider, Class<E> type, String name) {
        super(type, new EnumConverter<>(type), provider, name, null);
    }

    public CTPrefEnumProperty(@NotNull CTPreferencesProvider provider, Class<E> type, String name, E def) {
        super(type, new EnumConverter<>(type), provider, name, def);
    }

    public CTPrefEnumProperty(@NotNull CTPreferencesProvider provider, @NotNull Class<E> type) {
        super(type, new EnumConverter<>(type), provider, type.getSimpleName(), null);
    }

    public CTPrefEnumProperty(@NotNull CTPreferencesProvider provider, @NotNull Class<E> type, E def) {
        super(type, new EnumConverter<>(type), provider, type.getSimpleName(), def);
    }
}

File: src/main/kotlin/gargoyle/ct/pref/impl/prop/CTPrefEnumProperty.kt
package gargoyle.ct.pref.impl.prop

import gargoyle.ct.convert.impl.EnumConverter
import gargoyle.ct.pref.CTPreferencesProvider
import kotlin.reflect.KClass

class CTPrefEnumProperty<E : Enum<E>> : CTPrefProperty<E> {
    constructor(provider: CTPreferencesProvider, type: KClass<E>, name: String) : super(
        type,
        EnumConverter<E>(type),
        provider,
        name,
        null
    )

    constructor(provider: CTPreferencesProvider, type: KClass<E>, name: String, def: E) : super(
        type,
        EnumConverter<E>(type),
        provider,
        name,
        def
    )

    constructor(provider: CTPreferencesProvider, type: KClass<E>) : super(
        type,
        EnumConverter<E>(type),
        provider,
        type.qualifiedName!!.substringAfterLast('.'),
        null
    )

    constructor(provider: CTPreferencesProvider, type: KClass<E>, def: E) : super(
        type,
        EnumConverter<E>(type),
        provider,
        type.qualifiedName!!.substringAfterLast('.'),
        def
    )
}
